tests/validators.py

```diff
@@ -0,0 +1,85 @@
+from unittest import TestCase
+
+from adapters.validators import (
+    EMail, Length, Schema, Type, UnicodeNormalize, ValidationError,
+)
+
+
+class ValidatorsTest(TestCase):
+    def setUp(self):
+        class UsernameSchema(Schema):
+            inputs = {'username'}
+            outputs = {'username', 'username_normalized'}
+            validators = [
+                Type('username', str),
+                Length('username', min=4, max=100),
+                UnicodeNormalize('username', out='username_normalized'),
+            ]
+
+        class EMailSchema(Schema):
+            inputs = {'email'}
+            validators = [
+                EMail('email'),
+            ]
+
+        class RegistrationSchema(Schema):
+            depends = [UsernameSchema, EMailSchema]
+
+            def username_in_email(self):
+                if self.username_normalized not in self.email:
+                    raise ValidationError("E-mail must contain username", field='email')
```

Comment:
How do we want to handle cross-field validation where the error is not associated with a specific field?